core/unit_test/TestAtomicOperations.hpp

```diff
@@ -368,6 +368,63 @@ bool atomic_op_test(T old_val, T update) {
   return result == 0;
 }
 
+template <class T>
+constexpr T relative_error_threshold = T(1.0e-15);
+
+template <class Op, class T, class ExecSpace>
+bool atomic_op_test_rel(T old_val, T update) {
+  Kokkos::View<T[3], ExecSpace> op_data("op_data");
+  Kokkos::deep_copy(op_data, old_val);
+  int result = 0;
+  Kokkos::parallel_reduce(
+      Kokkos::RangePolicy<ExecSpace>(0, 1),
+      KOKKOS_LAMBDA(int, int& local_result) {
+        auto fetch_result =
+            Op::atomic_op(&op_data(0), &op_data(1), &op_data(2), update);
+        T expected_val = Op::op(old_val, update);
+        Kokkos::memory_fence();
+        if (expected_val == T(0)) {
+          if (fabs(op_data(0)) > relative_error_threshold<T>) local_result += 1;
+          if (fabs(op_data(1)) > relative_error_threshold<T>) local_result += 2;
+          if (fabs(op_data(2)) > relative_error_threshold<T>) local_result += 4;
+          if (fetch_result.first != old_val) local_result += 8;
+          if (fabs(fetch_result.second) > relative_error_threshold<T>)
+            local_result += 16;
+        } else {
+          if (fabs((op_data(0) - expected_val) / expected_val) >
+              relative_error_threshold<T>)
+            local_result += 1;
+          if (fabs((op_data(1) - expected_val) / expected_val) >
+              relative_error_threshold<T>)
+            local_result += 2;
+          if (fabs((op_data(2) - expected_val) / expected_val) >
+              relative_error_threshold<T>)
+            local_result += 4;
+          if (fetch_result.first != old_val) local_result += 8;
+          if (fabs((fetch_result.second - expected_val) / expected_val) >
+              relative_error_threshold<T>)
+            local_result += 16;
```

Comment:
Your diff obfuscates the change you suggested
```diff
@@ -1,3 +1,4 @@
+          using Kokkos::fabs;
           if (fabs(op_data(0)) > relative_error_threshold<T>) local_result += 1;
           if (fabs(op_data(1)) > relative_error_threshold<T>) local_result += 2;
           if (fabs(op_data(2)) > relative_error_threshold<T>) local_result += 4;
```
The using declaration would be needed in the else-branch of the if-condition.
I don't see this as something that should hold this PR.

Reply:
> Even though this is copied from the non-"rel" test
> 
> ```
>           if (fabs(op_data(0)) > relative_error_threshold<T>) local_result |= 1;
>           if (fabs(op_data(1)) > relative_error_threshold<T>) local_result |= 2;
>           if (fabs(op_data(2)) > relative_error_threshold<T>) local_result |= 4;
>           if (fetch_result.first != old_val) local_result |= 8;
>           if (fabs(fetch_result.second) > relative_error_threshold<T>)
>             local_result |= 16;
> ```
> 
> should be cleaner with respect to the bit checks below so that it would also work with more than one iteration and express intent better.
> 
> If you really only want one iteration, you could also use a `parallel_for` with `Kokkos::View<int, Kokkos::SharedSpace> result_view("result_view");` and directly write to it in the kernel.
> 
> Ultimately, it would be nice to combine the two `atomic_op_test_*` implementations (possibly by using the `if (expected_val == T(0))` branch with a 0 tolerance for `atomic_op_test`).

Code is readable enough.
I do not intend to refactor unless more reviewers ask for it.